pkg/azdiskschedulerextender/azdiskschedulerextender_test.go

```diff
@@ -0,0 +1,874 @@
+				klog.Errorf("handlePrioritizeRequest: Error decoding filter request: %v", err)
+				t.Fatal(err)
+			}
+
+			if !gotExpectedPrioritizeList(actualPrioritizeList, test.expectedPrioritizeResult) {
+				t.Errorf("Actual prioritize response (%s) does not equal expected response.", prioritizeResultRecorder.Body)
+			}
+		})
+	}
+}
+
+//TODO test only checks the repsonse code. add check for response body
+func TestFilterAndPrioritizeInRandomizedLargeCluster(t *testing.T) {
+	var nodes []v1.Node
+	var nodeNames []string
+
+	//TODO increase numberOfPodsToSchedule when changing the implemention to reuse goroutines
+	stressTestSetupParams := map[string][3]int{
+		"low":  [3]int{500, 5000, 1000},
+		"avg":  [3]int{1000, 10000, 1000},
+		"high": [3]int{5000, 30000, 1000},
+	}
+
+	//save original clients
+	savedAzVolumeAttachmentExtensionClient := azVolumeAttachmentExtensionClient
+	savedAzDriverNodeExtensionClient := azDriverNodeExtensionClient
+	defer func() {
+		azVolumeAttachmentExtensionClient = savedAzVolumeAttachmentExtensionClient
+		azDriverNodeExtensionClient = savedAzDriverNodeExtensionClient
+	}()
+
+	for _, setupParams := range stressTestSetupParams {
+		var tokens = make(chan struct{}, 20)
+		var wg sync.WaitGroup
+		var clusterNodes []v1alpha1Client.AzDriverNode
+		var clusterVolumes []v1alpha1Client.AzVolumeAttachment
+		numberOfClusterNodes := setupParams[0]
+		numberOfClusterVolumes := setupParams[1]
+		numberOfPodsToSchedule := setupParams[2]
+
+		// generate large number of nodes
+		for i := 0; i < numberOfClusterNodes; i++ {
+			nodeName := fmt.Sprintf("node%d", i)
+			clusterNodes = append(clusterNodes, getDriverNode(fmt.Sprintf("driverNode%d", i), ns, nodeName, true))
+			nodes = append(nodes, v1.Node{ObjectMeta: meta.ObjectMeta{Name: nodeName}})
+			nodeNames = append(nodeNames, nodeName)
+		}
+
+		// generate volumes and assign to nodes
+		for i := 0; i < numberOfClusterVolumes; i++ {
+			clusterVolumes = append(clusterVolumes, getVolumeAttachment(fmt.Sprintf("volumeAttachment%d", i), ns, fmt.Sprintf("vol%d", i), fmt.Sprintf("node%d", rand.Intn(5000)), "Ready"))
+		}
+
+		testClientSet := fakeClientSet.NewSimpleClientset(
+			&v1alpha1Client.AzVolumeAttachmentList{
+				Items: clusterVolumes,
+			},
+			&v1alpha1Client.AzDriverNodeList{
+				Items: clusterNodes,
+			})
+
+		// continue with fake clients
+		azVolumeAttachmentExtensionClient = testClientSet.DiskV1alpha1().AzVolumeAttachments(ns)
+		azDriverNodeExtensionClient = testClientSet.DiskV1alpha1().AzDriverNodes(ns)
+
+		var errorChan = make(chan error, numberOfPodsToSchedule)
+		for j := 0; j < numberOfPodsToSchedule; j++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				var testPodVolumes []v1.Volume
+				// randomly assign volumes to pod
+				podVolCount := rand.Intn(256)
+				for i := 0; i < podVolCount; i++ {
+					testPodVolumes = append(testPodVolumes, v1.Volume{Name: fmt.Sprintf("vol%d", rand.Intn(numberOfClusterVolumes))})
```

Comment:
Looking at the MacOS unit test failure, it was due to the test exceeding the timeout of 30s. Most of the goroutines appears to be waiting to acquire the global mutex on the shared source used by the `rand.Intn()` implementation. To avoid this bottleneck, consider using a unique `Source` for each goroutine, e.g.

```go
seed := make([]byte, 8)
_, err := crypt.rand.Bytes(seed)
// handle errors
rng := rand.New(rand.NewSource(int64(binary.LittleEndian.Uint64(b[:]))))

podVolCount := rng.Intn(256)
...
```